proposals/dnd-list.md

```diff
@@ -0,0 +1,144 @@
+# Drag and Drop List
+
+## Usage
+
+### Basic usage
+
+```js
+import * as React from 'react';
+import {StatefulList} from 'baseui/dnd-list';
+
+export default () =>
+  <StatefulList
+    initialState={{items: ['Item 1', 'Item 2', 'Item 3']}}
+    onChange={({newState, oldIndex, newIndex}) => {}}
+  />
+```
+
+### Advanced usage
+
+```js
+import * as React from 'react';
+import {
+  StatefulList,
+  StyledList,
+  StyledItem
+} from 'baseui/dnd-list';
+import {styled} from 'baseui';
+
+CustomList = styled(StyledList, props => ({
+  border: '1px solid green'
+}));
+
+CustomItem = styled(StyledItem, props => ({
+  borderBottom: props.$dragged ? `1px solid red` : 'none'
+}));
+
+export default () =>
+  <StatefulList
+    initialState={{ items: ['Item 1', 'Item 2', 'Item 3']}}
+    overrides={{
+      List: CustomList,
+      Item: CustomItem
+    }}
+  >
+```
+
+## Exports
+
+* `StatefulList`
+* `List`
+* `StyledList`
+* `StyledItem`
+* `StyledDragHandle`
+* `StyledCloseHandle`
+* `STATE_CHANGE_TYPE`
+
+## `List` API
+
+* `items: Array<string>` - Required.
+  List items.
+* `overrides: {Root, Item, DragHandle, CloseHandle}` - Optional.
+  Overrides for presentational components. See "Presentational Components Props API" below
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+* `onChange: ({oldIndex: number, newIndex: number})` - Optional.
+  onChange handler is called every time an item is moved or removed (newIndex = -1)
+
+## `StatefulList` API
+
+* `initialState: {items: Array<string>}` - Required.
+  Initial state of the component where `items` are list items.
+* `overrides: {Root, List, Item, DragHandle, CloseHandle}` - Optional.
+  Overrides for presentational components. See "Presentational Components Props API" below
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+* `onChange: ({newState: Array<string>, oldIndex: number, newIndex: number})` - Optional.
+  onChange handler that is called every time an item is moved or removed (newIndex = -1).
+* `stateReducer: (type: string, oldIndex: number, newIndex: number, currentState: {items: Array<string>}) => stateToSet: {items: Array<string>}` - Optional. Useful when you want to replace default array-move logic.
```

Comment:
We should probably stick to the signature the rest of components have and add additional args last `stateReducer(type, nextState, currentState)`.

Reply:
Good point! My thought was that it's pretty hard to tell what exactly changed if you have to compare two arrays and don't have `oldIndex` / `newIndex` information. It can be even impossible if the two items are not unique (and they don't have to be).

**First, I added `onSort` prop with this signature:** 

```tsx
(currentState: Array<string>, oldIndex: number, newIndex: number) : Array<string>
```

since I think that it's a legit requirement to let users override the sorting logic. For example there can be some extra validation allowing / disallowing particular drag and drop.

**But then I saw the `stateReducer`** and thought it doesn't make sense to have this very similar hook twice. So I removed `onSort` and modified `stateReducer`. But I can add `onSort` back and assume it will be applied first before `stateReducer` is triggered?

Reply:
Since we already provide `onChange` hook, for the stateful list `onChange`+`stateReducer` should be sufficient. And for further control, a stateless list should be used. Does that sound reasonable?

Reply:
It does! Updated.